### ISSUE TITLE
Import error

### ISSUE THREAD

**Issue by CiernaOvca** [AngeloDamiani/Ryu_RSTP#1]
Hello.

We are trying to run your RSTP switch, but it always throws an error: 

mininet@mininet-vm:~/ryu/ryu/app$ ryu-manager simple_switch_rstp.py
loading app simple_switch_rstp.py
Traceback (most recent call last):
  File "/usr/local/bin/ryu-manager", line 9, in <module>
    load_entry_point('ryu==4.17', 'console_scripts', 'ryu-manager')()
  File "/usr/local/lib/python2.7/dist-packages/ryu/cmd/manager.py", line 98, in main
    app_mgr.load_apps(app_lists)
  File "/usr/local/lib/python2.7/dist-packages/ryu/base/app_manager.py", line 415, in load_apps
    cls = self.load_app(app_cls_name)
  File "/usr/local/lib/python2.7/dist-packages/ryu/base/app_manager.py", line 392, in load_app
    mod = utils.import_module(name)
  File "/usr/local/lib/python2.7/dist-packages/ryu/utils.py", line 101, in import_module
    return _import_module_file(modname)
  File "/usr/local/lib/python2.7/dist-packages/ryu/utils.py", line 90, in _import_module_file
    return load_source(modname, abspath)
  File "/usr/local/lib/python2.7/dist-packages/ryu/utils.py", line 41, in load_source
    return imp.load_source(name, pathname)
  File "/home/mininet/ryu/ryu/app/simple_switch_rstp.py", line 9, in <module>
    from ryu.lib import rstplib
ImportError: cannot import name rstplib

rstplib.py is already in ryu/ryu/lib directory. 

Can you help us with this problem? 

Thank you in advance!

**Comment by AngeloDamiani** [comment]
Hello,
Do you have rstplib.py in /usr/local/lib/python2.7/dist-packages/ryu/lib directory?

As I can see from your terminal text, I see that you have installed Ryu using pip (or maybe you have created an alias for "ryu-manager" but I don't think so) but you have put rstplib.py in the lib directory of the source code that you have downloaded and put into your home directory.
So if you look for the ryu installation with pip (the path that I've written before) you can find the original lib folder. ryu-manager will be looking for rstplib.py there.

I hope that this can help

**Comment by CiernaOvca** [comment]
Hello! 

Thank you very much, it helped. But another problem has occured - when I have topology, which consists of 2 switches, everything works fine. But when there are more switches, ping does not work (Destination host unreachable). 

Here is my topology: 
from mininet.topo import Topo

class MyTopo( Topo ):
    "Simple topology example."

    def __init__( self ):
        "Create custom topo."

        # Initialize topology
        Topo.__init__( self )

        # Add hosts and switches
        h1 = self.addHost( 'h1' )
        h2 = self.addHost( 'h2' )
        h3 = self.addHost( 'h3' )
        h4 = self.addHost( 'h4' )
		
        s1 = self.addSwitch( 's1' )
        s2 = self.addSwitch( 's2' )
        s3 = self.addSwitch( 's3' )

        # Add links
        self.addLink( h1, s1 )
        self.addLink( h2, s2 )
        self.addLink( h3, s3 )
		
        self.addLink( s1, s2 )
        self.addLink( s2, s3 )


topos = { 'mytopo': ( lambda: MyTopo() ) }

And here is topology, that works: 
from mininet.topo import Topo

class MyTopo( Topo ):
    "Simple topology example."

    def __init__( self ):
        "Create custom topo."

        # Initialize topology
        Topo.__init__( self )

        # Add hosts and switches
        h1 = self.addHost( 'h1' )
        h2 = self.addHost( 'h2' )
        h3 = self.addHost( 'h3' )
		
        s1 = self.addSwitch( 's1' )
        s2 = self.addSwitch( 's2' )

        # Add links
        self.addLink( h1, s1 )
        self.addLink( h2, s2 )
		
        self.addLink( s1, s2 )



topos = { 'mytopo': ( lambda: MyTopo() ) }

I'm sorry for bothering you, but I don't know, what to do with this. Should I enable rstp on every switch in topology (with  ovs-vsctl1 ovs-vsctl set s0 rstp-enable=true)? 

Thanks in advance for your time and answer!

**Comment by AngeloDamiani** [comment]
Please, can you show me the output of the "good" topology? Tomorrow I'll see and I'll try to help

**Comment by CiernaOvca** [comment]
![image](https://user-images.githubusercontent.com/11999506/33171210-aea164d2-d04b-11e7-8afe-8264c1c4ef39.png)

There you can see that ping works fine for two switches. 

You can see in topology (on right side, that there are some commented lines. If I uncommend these lines and try ping from h1 to h3, it won't work (destination host unreachable). I really don't know, what I'm doing wrong. 

Thanks in advance!

**Comment by AngeloDamiani** [comment]
Can you show me the same two windows for the "bad" topology?

**Comment by CiernaOvca** [comment]
![image](https://user-images.githubusercontent.com/11999506/33174533-c91a47a4-d058-11e7-8a78-023e6d386e56.png)

here it is

**Comment by AngeloDamiani** [comment]
Sincerely I can't understand why is not working properly. Let's try in another way. Create a python file (call it topo_rstp.py) and copy and paste the following code:

`#!/usr/bin/python

from mininet.net import Mininet
from mininet.node import Controller, RemoteController, OVSController
from mininet.node import CPULimitedHost, Host, Node
from mininet.node import OVSKernelSwitch, UserSwitch
from mininet.node import IVSSwitch
from mininet.cli import CLI
from mininet.log import setLogLevel, info
from mininet.link import TCLink, Intf
from subprocess import call

def myNetwork():

    net = Mininet(ipBase='10.0.0.0/8')

    info( '*** Adding controller\n' )
    c0=net.addController(name='c0',
                      controller=RemoteController,
                      protocol='tcp',
                      port=6653)

    info( '*** Add switches\n')
    s1 = net.addSwitch('s1', protocols="OpenFlow13")
    s2 = net.addSwitch('s2', protocols="OpenFlow13")
    s3 = net.addSwitch('s3', protocols="OpenFlow13")

    info( '*** Add hosts\n')
    h1 = net.addHost('h1')
    h2 = net.addHost('h2')
    h3 = net.addHost('h3')

    info( '*** Add links\n')
    net.addLink(s1, s2,1,1)          #Create a link between port 1 of s1 and port 1 of s2
    net.addLink(s2, s3,2,1)          #Create a link between port 2 of s2 and port 1 of s3

    net.addLink(s1, h1)
    net.addLink(s2, h2)
    net.addLink(s3, h3)

    net.build()
    for controller in net.controllers:
        controller.start()

    net.get('s1').start([c0])
    net.get('s2').start([c0])
    net.get('s3').start([c0])

    CLI(net)
    net.stop()

if __name__ == '__main__':
    setLogLevel( 'info' )
    myNetwork()

`

And start it using: sudo python topo_rstp.py .
I tried this one and it works; tell me if it doesn't work. I'll wait for your feedback

**Comment by CiernaOvca** [comment]
Hello there! 

Thank you very much, it works! But I'm not sure why.. do you think, that it is because of the ports? Should I always define, on which port I want to connect? Or is it because of controller? 

I'm also not familiar with: 
for controller in net.controllers:
    controller.start()

But thank you very much for your time! I definitely owe you a beer, if you ever find yourself in Slovakia!

**Comment by CiernaOvca** [comment]
![image](https://user-images.githubusercontent.com/11999506/33218743-d6287876-d13e-11e7-99a5-59f64930ea00.png)
 
Hello,

I tried it with yours topology and it worked. But when I have changed it to a topology I need, it fails. I thought, it was because I need a ring topology which your rstp might don't support, but even when I removed the last link it still wouldn't work. I really don't know where is the problem. 

Maybe we have different versions of some component or something? I'm using mininet 2.2.2 and OVS 2.5.0.

**Comment by CiernaOvca** [comment]
![image](https://user-images.githubusercontent.com/11999506/33218980-2a075c72-d140-11e7-9055-6590cd3e6ab0.png)

Here is my topology.

**Comment by AngeloDamiani** [comment]
I tried your last topology and it works for me.

I just had to do a little change. I had to add the IP address of the machine in which I'm running the controller (because i'm running mininet in virtual machine but I'm running the controller directly on my pc). So I added the line `ip="192.168.56.1"` into the addController(...) parameters.

By the way I'm currently using  OVS 2.0.2, mininet 2.2.2. Maybe is for OVS's version but It's really strange.

That's my output (i voluntarily cut the packets flows from the output in order to let you see the RSTP convergence):

![screenshot from 2017-11-25 13-36-09](https://user-images.githubusercontent.com/12428839/33230556-fad9dd70-d1e5-11e7-841d-8b2614fe13bd.png)

and that's my topology code:

`
def myNetwork():

    net = Mininet(ipBase='10.0.0.0/8')

    info( '*** Adding controller\n' )
    c0=net.addController(name='c0',
                      controller=RemoteController,
                      ip='192.168.56.1',
                      protocol='tcp',
                      port=6653)

    info( '*** Add switches\n')
    s0 = net.addSwitch('s0', protocols="OpenFlow13")
    s1 = net.addSwitch('s1', protocols="OpenFlow13")
    s2 = net.addSwitch('s2', protocols="OpenFlow13")
    s3 = net.addSwitch('s3', protocols="OpenFlow13")
    #s4 = net.addSwitch('s4', protocols="OpenFlow13")
    #s5 = net.addSwitch('s5', protocols="OpenFlow13")

    info( '*** Add hosts\n')
    h0 = net.addHost('h0')
    h1 = net.addHost('h1')
    h2 = net.addHost('h2')
    h3 = net.addHost('h3')
    h4 = net.addHost('h4')
    h5 = net.addHost('h5')
    h6 = net.addHost('h6')
    h7 = net.addHost('h7')
    h8 = net.addHost('h8')
    h9 = net.addHost('h9')
    info( '*** Add links\n')
    net.addLink(s0, s1,1,1)
    net.addLink(s1, s2,2,1)
    net.addLink(s2, s3,2,1)

    net.addLink(s0, h0)
    net.addLink(s0, h1)
    net.addLink(s0, h2)
    net.addLink(s1, h3)
    net.addLink(s1, h4)
    net.addLink(s2, h5)
    net.addLink(s2, h6)
    net.addLink(s2, h7)
    net.addLink(s3, h8)
    net.addLink(s3, h9)

    net.build()
    for controller in net.controllers:
        controller.start()

    net.get('s0').start([c0])
    net.get('s1').start([c0])
    net.get('s2').start([c0])
    net.get('s3').start([c0])

    CLI(net)
    net.stop()

if __name__ == '__main__':
    setLogLevel( 'info' )
    myNetwork()
`

By the way, my bachelor's degree final project was about developing this library (and this application) for ryu and testing it on a ring topology made of 5 switches. So hopefully this should work even in your ring case. 
In the end, my only advice is to try to run mininet with OVS 2.0.2.

And yes, I'll accept that beer if I will ever come to Slovakia :)

**Comment by CiernaOvca** [comment]
Hello!

I tried your topology, but again it didn't work, I really don't know what is wrong. I changed the controller IP to virtual machine's IP (I run mininet and then ifconfig command) and then tried your topology file, but there's still 'Destination host unreachable'. 

I also tried to change the OVS version, but linux kernel of my virtual machine does not support it.

I have two other projects, which I have to finish, so I'm giving up. I will run rstp manually on switches, I hope it won't had major impact on my grade. 

Anyway, thank you for you time and patience, and good luck with your other amazing projects!
